rfd/0182-multi-port-tcp-app-access.md

```diff
@@ -0,0 +1,496 @@
+```
+
+##### Error path
+
+###### Incorrect port
+
+Dave wants to connect to the debug port through VNet, but he makes a typo and provides a port
+number that's not in the app spec.
+
+```
+waldo-debug waldo.teleport.example.com:5096
+```
+
+The connection reaches the app service which closes the connection because of an unknown port.
+waldo-debug observes the connection being unexpectedly closed. Dave doesn't see any immediate
+information about an incorrect port – the app service has no way of passing this information as it
+does not know about the protocol used by waldo. Dave has to ask Alice to look into the logs of the
+app agent. She notices that around the time when the connection was made, the app service logged a
+warning saying that port 5096 is not defined in the app spec.
+
+```
+2024-08-23T16:47:02+02:00 WARN [APP:SERVI] Failed to handle client connection. error:[
+ERROR REPORT:
+Original Error: *trace.BadParameterError port 5096 not found in ports of the app "waldo"
+Stack Trace: …
+```
+
+###### Invalid app spec
+
+Alice wants to extend the app spec with another port, 6060, but she makes a mistake and forgets a
+comma:
+
+```yaml
+apps:
+- name: "waldo"
+  uri: "tcp://localhost"
+  ports: [4080-4090, 50956060]
+  labels:
+    env: prod
+```
+
+The relevant UI (`teleport start`, `tctl edit`, etc.) returns an error about an incorrect port
+in the app spec and prevents Alice from saving such app spec.
+
+### Passing the port number from the client to the app agent
+
+In order to pass the port number from the client to the app service, the underlying local proxy is
+going to include the port number as a part of a special subdomain in the SNI, e.g.
+`app-teleport-proxy-target-port-1337.teleport.cluster.local`. The ALPN proxy is then able to extract
+the port number out of the SNI and pass it to the app service through a new field in
+`sshutils.DialReq`.
+
+When the port makes its way to the app service, the app service checks the port against ports
+defined in the app spec before deciding whether to proxy the connection or not. The app service
+[already has access to the app spec](https://github.com/gravitational/teleport/blob/5012036ad28a25d54858e454364828803f558e0b/lib/srv/app/connections_handler.go#L726).
+
+At this point we don't plan to introduce RBAC for port numbers, but this is something we can
+consider in the future. Port numbers on an individual app are akin to database names within a single
+database server. Customers might want to treat them as such, by e.g. limiting access to different
+ports.
+
+#### Advantages
+
+* A single cert per app.
+* With per-session MFA enabled, the user needs to tap an MFA device only once per app.
+* Old clients that don't send info about target port can still be routed to the first port found in
+  the `ports` field.
+* Backward-compatible with existing UIs (tsh, tctl, Web UI, Connect). Requires minimal effort to add
+  support for multi-port to those UIs.
+
+#### Disadvantages
```

Comment:
Another disadvantage: it probably breaks server name verification because the proxy's TLS certificate is unlikely to include this value as a valid hostname